internal/stacks/stackManager.go

```diff
@@ -0,0 +1,597 @@
+	for _, member := range s.Stack.Members {
+		if member.External {
+			configFilename := path.Join(constants.StacksDir, s.Stack.Name, "configs", fmt.Sprintf("firefly_core_%v.yml", member.ID))
+			var port int
+			if firstTimeSetup {
+				port = member.ExposedFireflyAdminPort
+			} else {
+				port = member.ExposedFireflyPort
+			}
+			// Check process running
+			available, err := checkPortAvailable(port)
+			if err != nil {
+				return err
+			}
+			if available {
+				s.Log.Info(fmt.Sprintf("please start your firefly core with the config file for this stack: firefly -f %s  ", configFilename))
+				if err := s.waitForFireflyStart(port); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (s *StackManager) waitForFireflyStart(port int) error {
+	retries := 120
+	retryPeriod := 1000 // ms
+	retriesRemaining := retries
+	for retriesRemaining > 0 {
+		time.Sleep(time.Duration(retryPeriod) * time.Millisecond)
+		available, err := checkPortAvailable(port)
+		if err != nil {
+			return err
+		}
+		if !available {
+			return nil
+		}
+		retriesRemaining--
+	}
+	return fmt.Errorf("waited for %v seconds for firefly to start on port %v but it was never available", retries*retryPeriod/1000, port)
+}
+
+func (s *StackManager) UpgradeStack(verbose bool) error {
+	workingDir := filepath.Join(constants.StacksDir, s.Stack.Name)
+	if err := docker.RunDockerComposeCommand(workingDir, verbose, verbose, "down"); err != nil {
+		return err
+	}
+	return docker.RunDockerComposeCommand(workingDir, verbose, verbose, "pull")
+}
+
+func (s *StackManager) PrintStackInfo(verbose bool) error {
+	workingDir := filepath.Join(constants.StacksDir, s.Stack.Name)
+	fmt.Print("\n")
+	if err := docker.RunDockerComposeCommand(workingDir, verbose, true, "images"); err != nil {
+		return err
+	}
+	fmt.Print("\n")
+	if err := docker.RunDockerComposeCommand(workingDir, verbose, true, "ps"); err != nil {
+		return err
+	}
+	fmt.Printf("\nYour docker compose file for this stack can be found at: %s\n\n", filepath.Join(constants.StacksDir, s.Stack.Name, "docker-compose.yml"))
+	return nil
+}
+
+func (s *StackManager) patchConfigAndRestartFireflyNodes(verbose bool) error {
+	for _, member := range s.Stack.Members {
+		s.Log.Info(fmt.Sprintf("applying configuration changes to %s", member.ID))
+		configRecordUrl := fmt.Sprintf("http://localhost:%d/admin/api/v1/config/records/admin", member.ExposedFireflyAdminPort)
+		if err := s.httpJSONWithRetry("PUT", configRecordUrl, "{\"preInit\": false}", nil); err != nil && err != io.EOF {
+			return err
+		}
+		resetUrl := fmt.Sprintf("http://localhost:%d/admin/api/v1/config/reset", member.ExposedFireflyAdminPort)
+		if err := s.httpJSONWithRetry("POST", resetUrl, "{}", nil); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *StackManager) StackHasRunBefore() (bool, error) {
+	path := filepath.Join(constants.StacksDir, s.Stack.Name, "data", fmt.Sprintf("dataexchange_%s", s.Stack.Members[0].ID), "cert.pem")
```

Comment:
We should probably come up with a more definitive way to do this (adding a value or file to the config, outside of dataexchange).

Reply:
But that's probably a note for a future enhancement (good-first-issue?)

Reply:
Agreed. Whatever the check is, the `reset` command will need to flip it back though.

Reply:
I'd say that can probably be handled outside the scope of this PR though. But I agree it's something we should improve.